File: Data Structures & Algorithms/generate-parentheses/submission-4.js
class Solution {
    /**
     * @param {number} n
     * @return {string[]}
     */
    generateParenthesis(n) {
        let res = [];
        const b = (n, open, close, cur, res) => {
            if (n === open && open === close) {
                res.push(cur)
                return
            }

            if (n > open) {
                b(n, open + 1, close, cur + '(', res)
            }

            if (open > close) {
                b(n, open, close + 1, cur + ')', res)
            }

        }
        b(n, 0, 0, '', res)
        return res
    }
}

File: Data Structures & Algorithms/open-the-lock/submission-4.py
class Solution:
    def openLock(self, deadends: List[str], target: str) -> int:
        q = deque(['0000']) 
        visited = set(['0000'])  
        d = set(deadends) 

        res = 0 

        while q: 
            for _ in range(len(q)): 
                cur = q.popleft() 
                if cur == target: 
                    return res  

                if cur in d: 
                    continue

                for i in range(8): 
                    position = i // 2 
                    direction = i % 2  
                    number = int(cur[position])  
                    next = list(cur)

                    if direction == 0:  
                        if number == 0: 
                            number = 9 
                        else: 
                            number -= 1   
                    else: 
                        if number == 9: 
                            number = 0 
                        else: 
                            number += 1 
                           
                    next[position] = str(number)  
                    next = "".join(next)

                    if next not in visited:  
                        q.append(next) 
                        visited.add(next)  

            res += 1 

        return -1
                        